### DISCUSSION TITLE
Run CasaOS inside a container

### DISCUSSION THREAD

**Discussion by bram-nijenkamp** [IceWhaleTech/CasaOS#217] [Ideas]
**Is your feature request related to a problem? Please describe.**
I Would like to try CasaOS; but wouldn't it be possible to run in from inside a ubuntu based docker container?

**Describe the solution you'd like**
Is it an option to just create a docker file with ubuntu and docker installed; install it through the sh file and then pass the docker socket from the host. Has anybody yet tried this?

**Describe alternatives you've considered**
-

**Additional context**
-

**Comment by JohnGuan** [comment]
Theoretically, it should be possible. but we have no plans to make it a Docker container yet.

However, CasaOS is still in its early stages, you can notice that we have "OS" in our name but it is still only installed as a system service in base OS. Because we have a lot of features in the development plan.

But I would love to know why you would like CasaOS to be installed in a container and what features you would like to see in this form.

**Reply by Martiiim** [reply]
well i almost did it but i didnt bcz of some problem whit the docker daemon
im gona explain so i ran a ubuntu docker image but whit a maximum of 1g so i can install casa os
then i did the following comands `apt-update && apt-upgrade` then `curl -fsSL https://get.casaos.io | bash` but then something went wrong whit the docker daemon

if anyone solves it pls send the solution

**Reply by mhagnumdw** [reply]
> Theoretically, it should be possible. but we have no plans to make it a Docker container yet.

@JohnGuan , still out of the plans??

**Comment by ddproxy** [comment]
Could, theoretically, run the majority of the stack within a docker stack. Having it this way would enable running it alongside an existing docker cluster or installation without meddling with the host system - and designing the system to be container-based from the get-go may improve portability of the system (to different OS's or to a multi-host like kubernetes installation).

Just, as a concept. I haven't dug into how CasaOS works fundamentally, yet.

**Comment by tigerinus** [comment]
It'd be nice to consider [Docker-in-Docker (DIND)](https://hub.docker.com/_/docker/) or K8s in Docker (KIND) mode such that all CasaOS' containers are wrapped inside a CasaOS container context.

OR,

consider containerd level containerization for CasaOS as a grouping. Basically you run `nerdctl ps` to see all containerd based containers, and run `docker ps` to see all docker based containers. 

Either approach above quarantines all CasaOS stuff together so user have a cleaner docker environment for their own use.

I personally prefer first approach (DIND/KIND) so that I can easily spawn a CasaOS instance anywhere without worrying too much about the level of intrusion from CasaOS to my host OS. I can also easy disable CasaOS by simply docker stop or kuberctl scale --replica=0 in one step, and vice versa for enabling. Not to mention it's also much easier for CasaOS dev team to test with DIND/KIND due to the decoupling.

Another benefit of DIND is it make it easier to run CasaOS on Windows with Docker or Mac with Docker. People might run Windows on ZB. A large number of HTPC market is also Windows based. CasaOS should bring the app-ized home server apps to all platforms.

**Comment by LouvAndTech** [comment]
It would be extremely pleasant to use it alongside an already existing homelab setup.
First of all as a container manager (Almost like a portainer) and also as a well thought interface (I love it btw) for a whole home setup.
The ability to deploy stack would be useful too.

**Comment by Hedius** [comment]
Hi i have a finished docker image for casaos which works fully. still needs some polishing, but I will publish in 1-2 weeks when i have time.

**Reply by axute** [reply]
Can you provide the docker image? I would be very interested.

**Reply by phoenixsfly** [reply]
+1

**Reply by Hedius** [reply]
The new casaos version needs a different layout. My image no longer works at all.

I stopped using casaos. Since i am a pure console user. And I lack the time sadly to maintain this. Someone else will have to do this sorry :)

**Reply by axute** [reply]
That's a pity, I looked at it, it doesn't look easy. A lot of dependencies.

**Reply by tigerinus** [reply]
It might be possible with https://github.com/nestybox/sysbox

**Comment by wjcroom** [comment]
I run it in  PVE8.0   LXC debian12 just chang sourc.list， will  install  fast，  contain  option0 fuse  open

**Reply by Jarusz** [reply]
Could u give more detailed instructions? As someone who uses docker-compose mainly from linuxserver and such i have no idea what to do from this, and id love to run casaos in a docker container...

**Comment by Metazla** [comment]
Hi, I just finished packing CasaOS in a Docker image. I will mainly use it as a nicer Portainer for my NAS and thought it would interest someone here so I am sharing.
https://github.com/Metazla/casa-img

**Reply by SfirreFan** [reply]
Super, I want to install casaos on my synology. I have no idea how to install your image. Do you have a discription how to do it? And wil the nas be fast enough?

**Reply by Metazla** [reply]
I updated the README to provide instructions👍. You will have to build the image since it is not available on DockerHub yet. I think your NAS should handle it since CasaOS is lightweight on resources (RAM and CPU), and Docker doesn't significantly increase resource usage. The main issue I foresee on Synology is the architecture, as I believe Synology uses ARM. I use the image on my x86 NAS, so I haven't tried it on ARM.
